### PR TITLE
Discussions | OG Card preview and data in the editors

### PR DESCRIPTION
**Pull request by bkoval**
## Needs
#2443 #2362 

## Links

* http://wikia-inc.atlassian.net/browse/SOC-2237
* http://wikia-inc.atlassian.net/browse/SOC-1936
* http://wikia-inc.atlassian.net/browse/SOC-1386
* http://wikia-inc.atlassian.net/browse/SOC-1937

## Description

* Styling OG Card in the post / reply / edit editors (based on styles in SOC-1940)
* Ability to paste or type in the url for OG lookup
* Associating OG data with the post after saving
* Introducing auto-sizeable textarea inside the editor

## Reviewers
@rafalkalinski @kvas-damian

### CODE REVIEW

File: front/main/app/components/discussion-edit-editor.js [hunk 1]
```diff
@@ -79,13 +79,27 @@ export default DiscussionEditorComponent.extend({
 		}
 	},
 
+	setOpenGraphProperties(text, urlRegex) {
+		if (this.get('discussionEditor.discussionEntity.rawContent') === this.get('bodyText')) {
```
Comment (kvas-damian) [inline front/main/app/components/discussion-edit-editor.js]: This solution seems to not be bulletproof. Please try to think of a solution that is less hacky.

File: front/main/app/components/discussion-editor.js [hunk 2]
```diff
@@ -6,17 +6,21 @@ export default Ember.Component.extend(ViewportMixin, {
 	attributeBindings: ['style'],
 
 	classNames: ['discussion-editor'],
-	classNameBindings: ['isActive'],
+	classNameBindings: ['isActive', 'hasError', 'showsOpenGraphCard:has-open-graph'],
```
Comment (kvas-damian) [inline front/main/app/components/discussion-editor.js]: why do we need `hasError`?

File: front/main/app/mixins/discussion-contribution-route.js [hunk 3]
```diff
@@ -122,5 +122,14 @@ export default Ember.Mixin.create({
 				this.get('discussionEditEditor').set('isLoading', false);
 			});
 		},
+
+		generateOpenGraph(uri) {
+			const model = this.modelFor(this.get('routeName'));
+
+			return model.generateOpenGraph(uri).then((openGraph) => {
+				// TODO do something with the data
```
Comment (kvas-damian) [inline front/main/app/mixins/discussion-contribution-route.js]: Can we somehow resolve this TODO?

File: front/main/app/styles/component/_discussion-editor.scss [hunk 4]
```diff
@@ -227,18 +232,19 @@ $editor-textarea-overflow-ios-hack: 30px;
 		}
 
 		.editor-wrapper {
-			height: calc(100% - #{$editor-toolbar-height});
+			height: calc(100% - #{2 * $editor-toolbar-height});
```
Comment (kvas-damian) [inline front/main/app/styles/component/_discussion-editor.scss]: Probably we should use `label-height` and `$editor-toolbar-height`. They're not equivalent.

File: front/main/app/components/discussion-editor.js [hunk 5]
```diff
@@ -217,21 +275,72 @@ export default Ember.Component.extend(ViewportMixin, {
...
+	 * that's why there's a need to wait a bit to perform regex lookup for the url
+	 *
+	 * @param {Event} event
+	 *
+	 * @returns {void}
+	 */
+	onPaste(event) {
+		Ember.run.later(() => {
+			this.setOpenGraphProperties(event.target.value, /(https?:\/\/[^\s]+)/g);
+		}, 100);
```
Comment (kvas-damian) [inline front/main/app/components/discussion-editor.js]: Is it possible to get rid of `setTimeout`?

File: front/main/app/templates/components/discussion-open-graph.hbs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+{{/if}}
+<div class="og-texts">
+	<span class="og-title">
+		{{openGraphData.title}}
+	</span>
+	<span class="og-site-name">
+		{{siteName}}
+	</span>
+</div>
+{{loading-spinner
```
Comment (kvas-damian) [inline front/main/app/templates/components/discussion-open-graph.hbs]: What do you think about wrapping this component in an `if`? This way we can avoid adding unnecessary HTML markup when it's not required.

File: front/main/app/styles/component/_discussion-editor.scss [hunk 7]
```diff
@@ -227,18 +232,20 @@ $editor-textarea-overflow-ios-hack: 30px;
 		}
 
 		.editor-wrapper {
-			height: calc(100% - #{$editor-toolbar-height});
+			//editor label + editor toolbar
+			height: calc(100% - #{36px + $editor-toolbar-height});
```
Comment (kvas-damian) [inline front/main/app/styles/component/_discussion-editor.scss]: Maybe we should extract `36px` into a variable?

File: front/main/app/components/discussion-editor.js [hunk 8]
```diff
@@ -217,21 +277,72 @@ export default Ember.Component.extend(ViewportMixin, {
...
 		this._super(...arguments);
-
+		this.initializePasting();
 		this.handleIOSFocus();
 		this.initializeStickyState();
+		this.initializeSizing();
+	},
+
+	initializePasting() {
+		this.$().find('textarea').on('paste', $.proxy(this.onPaste, this));
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-editor.js]: this.$.proxy

File: front/main/app/components/discussion-editor.js [hunk 9]
```diff
@@ -217,21 +277,72 @@ export default Ember.Component.extend(ViewportMixin, {
...
+	},
+
+	/**
+	 * Sets textarea height based on it's scrollHeight
+	 *
+	 * @returns {void}
+	 */
+	setSize() {
+		this.style.height = '1px';
+		this.style.height = `${this.scrollHeight}px`;
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-editor.js]: We're using jQuery for that kind of manipulation it all other places

Comment (bkoval) [inline front/main/app/components/discussion-editor.js]: I know, but here a context is a Javascript DOM node object passed by jQuery itself. Do you think it makes sense to wrap it in $() again here for just 2 simple operations?

File: front/main/app/templates/components/discussion-open-graph.hbs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+{{/if}}
+<div class="og-texts">
+	<span class="og-title">
+		{{openGraphData.title}}
+	</span>
+	<span class="og-site-name">
+		{{siteName}}
+	</span>
+</div>
+{{#if hasLoadingState}}
```
Comment (kvas-damian) [inline front/main/app/templates/components/discussion-open-graph.hbs]: Why can't we just use `isLoading`?

**Comment by kvas-damian** [comment]
LGTM

**Comment by rafalkalinski** [comment]
LGTM

**Pull request merged**